Pyroclastic anomaly tweak (anti-box)

0x6273 opened the pull request:
Pyroclastic anomaly will now get more unstable at very high temperatures, so if you box it in you will need a freezer to avoid it going critical.

The reason for this change is that before, you could just box it in with glass and not have to deal with it. This is quite boring and Emo has mentioned wanting to change this. See PR #17293.

It's implemented with a `TempReactingAnomalyComponent` that is reusable for other anomalies.

:cl:
- tweak: The pyroclastic anomaly will become more unstable at extreme temperatures, such as when confined to a small space.

deltanedas commented:
now you just box it and space the floor

EmoGarbage404 commented:
> now you just box it and space the floor

it has a min and max temp, so you need to manage it both ways.

deltanedas commented:
ok then scrub plasma into space easy

metalgearsloth commented on Content.Server/Anomaly/Components/TempReactingAnomalyComponent.cs:
```diff
@@ -0,0 +1,63 @@
+    /// <summary>
+    /// A list of reactions that happen at certain temperatures.
+    /// </summary>
+    [DataField("reactions")]
+    public List<TempReaction> Reactions = new();
```
Should this just be an includedatadef instead.

0x6273 replied:
Assuming you mean `IncludeDataField` this does not seem to work with `List`. I just changed it to be `required: true`.

metalgearsloth commented on Content.Server/Anomaly/Effects/TempReactingAnomalySystem.cs:
```diff
@@ -0,0 +1,45 @@
+        if (mixture is null)
+            return;
+
+        foreach (var reaction in tempReactingAnomaly.Reactions)
+        {
+            if (reaction.TempInRange(mixture.Temperature) && TryComp<AnomalyComponent>(uid, out var anomaly))
+            {
```
Do the TryComp out of the loop as it's not needed every iteration.

0x6273 commented:
Addressed review, test fail is unrelated to my changes since it happens on master: https://github.com/space-wizards/space-station-14/actions/runs/5287876287/jobs/9568895250

EmoGarbage404 commented:
By now, the pyroclastic anomaly is in a fine place balance-wise.

This change is unintuitive and not really in line with the other anomalies. I don't feel like this is the way to go about doing this.

AJCM-git commented:
Another better would probably be making high pressure/temperature always damage windows.

0x6273 commented:
> Another better would probably be making high pressure/temperature always damage windows.

I like this idea, but I think people will just use walls if this is only for windows. I guess those could also be damaged if temps get high enough to melt steel. I'm not sure how to code this in a good way since I imagine having every wall/window in the game check surrounding temps would probably tank performance.

AJCM-git commented:
> I like this idea, but I think people will just use walls if this is only for windows. I guess those could also be damaged if temps get high enough to melt steel. I'm not sure how to code this in a good way since I imagine having every wall/window in the game check surrounding temps would probably tank performance.

Well if they box with walls then they cant blast it with particles to avoid a meltdown. Tho pressure should definitively damage walls, idk enough about atmos code to tell you a way to do this efficiently, the only lead i would have for something like that is taking a look at how atmos superconductivity works (its an atmos CVAR that makes heat transportable through walls which **may** or may not help you arrive at a solution)
